scipy/interpolate/src/_dierckxmodule.cc

```diff
@@ -0,0 +1,775 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include <iostream>
+#include <string>
+#include <memory>
+#include "numpy/arrayobject.h"
+#include "__fitpack.h"
+
+
+/*
+ * Check obj to be an ndim C contiguous array of specified typenum
+ *
+ *  sort of mimic checks cython does for `double[::1] obj`
+ */
+static int
+check_array(PyObject *obj, npy_intp ndim, int typenum) {
+
+    int cond = (PyArray_CheckExact(obj) &&
+               (PyArray_TYPE((PyArrayObject*)obj) == typenum) &&
+               (PyArray_NDIM((PyArrayObject*)obj) == ndim) &&
+                PyArray_CHKFLAGS((PyArrayObject*)obj, NPY_ARRAY_ALIGNED | NPY_ARRAY_C_CONTIGUOUS)
+    );
+
+    if(!cond) {
+        std::string msg = ("Expected a " + std::to_string(ndim) + "-dim C contiguous array " +
+                           " of dtype = " + std::to_string(typenum) + "( got " +
+                           std::to_string(PyArray_TYPE((PyArrayObject*)obj)) +" )\n");
+        // XXX: name the dtype from typenum? Also type of arg if not array
+        PyErr_SetString(PyExc_ValueError, msg.c_str());
+        return 0;
+    }
+    return 1;
+}
+
+
+/*
+ * RAII malloc-d work arrays
+ */
+struct malloc_deleter {
+    void operator()(void *p) const { std::free(p); }
```

Comment:
I know. There is an internal function, `_de_Boor_D`, which expects a raw pointer to pre-allocated arena, and this was a way to `malloc` the space and have it deallocated automagically. Of course, this is a recipe from somewhere in the internet :-).

Anyhow, this is now gone in favor of just creating a `vector` in a function scope. The sequence is 

```
{
   std::vector<double> wrk;
   wrk.resize(2*k+2);

   call_low_level_stuff(...., wrk.data());

  ...

}  // wrk goes out of scope and is free-d
```

Reply:
Your example can be simplified using constructor (4) from [cppreference/vector](https://en.cppreference.com/w/cpp/container/vector/vector).

```cpp
{
   std::vector<double> wrk(2*k+2);

   call_low_level_stuff(...., wrk.data());

  ...

}  // wrk goes out of scope and is free-d
```

This will create an array of size ``2*k + 2`` filled with whatever [std::allocator_traits](http://en.cppreference.com/w/cpp/memory/allocator_traits)<A>::construct(m, p); returns for ``double``